runtime/gc_vlhgc/CopyForwardScheme.cpp

```diff
@@ -897,30 +920,48 @@ MM_CopyForwardScheme::reserveMemoryForObject(MM_EnvironmentVLHGC *env, UDATA com
 	/* 
 	 * 2. attempt to acquire a region from the tail candidates list 
 	 */
-	if ((NULL == result) && (NULL != _reservedRegionList[compactGroup]._tailCandidates)) {
-		_reservedRegionList[compactGroup]._tailCandidatesLock.acquire();
-		region = _reservedRegionList[compactGroup]._tailCandidates;
+	if ((NULL == result) && (NULL != _reservedRegionList[compactGroup]._freeMemoryCandidates)) {
+		_reservedRegionList[compactGroup]._freeMemoryCandidatesLock.acquire();
+		region = _reservedRegionList[compactGroup]._freeMemoryCandidates;
 		MM_HeapRegionDescriptorVLHGC *resultRegion = NULL;
 		while ((NULL == result) && (NULL != region)) {
-			MM_MemoryPoolBumpPointer *memoryPool = (MM_MemoryPoolBumpPointer*)region->getMemoryPool();
+			MM_MemoryPoolAddressOrderedList *memoryPool = (MM_MemoryPoolAddressOrderedList*)region->getMemoryPool();
 			Assert_MM_true(NULL != memoryPool);
 
 			/* make sure that we can't be copying objects into the area covered by a card which is meant to describe objects which were already in the region */
-			UDATA lostToAlignment = alignMemoryPool(env, memoryPool);
+			UDATA lostToAlignment = alignMemoryPool4Collector(env, memoryPool);
+
+			PORT_ACCESS_FROM_ENVIRONMENT(env);
+			j9tty_printf(PORTLIB, "alignMemoryPool4Collector lostToAlignment=%zu, region=%p, memoryPool=%p, _heapFreeList=%p, ActualFreeMemorySize=%zu, ActualFreeEntryCount=%zu, DarkMatterBytes=%zu\n",
+					lostToAlignment, region, memoryPool, memoryPool->getFirstFreeStartingAddr(env), memoryPool->getActualFreeMemorySize(), memoryPool->getActualFreeEntryCount(), memoryPool->getDarkMatterBytes());
+
 			env->_copyForwardCompactGroups[compactGroup]._discardedBytes += lostToAlignment;
+			if ((0 == memoryPool->getActualFreeEntryCount()) || (0 == memoryPool->getActualFreeMemorySize())) {
+
```

Comment:
I'm guessing this path is not necessary, but you want to avoid converting this region into survivor one, and only then realize it cannot satisfy reservation of object/copycache?

Reply:
can it actually ever be that count is non 0 and size is 0 or vice versa? in other words could we just check one and assert the other

Reply:
modified to check one and assert the another.